### PR TITLE
this commit resolve second part of #12862

### PR DESCRIPTION
**Pull request by rahulsangwn**
Improved rax2
some testing commands.
`rax2 -S "hello\x2Aworld ZYX\234abc\n"`
`rax2 -S "ra\233"`

Please review this commit

### CODE REVIEW

File: binr/rax2/rax2.c [hunk 1]
```diff
@@ -18,6 +18,7 @@ static int use_stdin();
 static int force_mode = 0;
 static int rax(char *str, int len, int last);
 static const char *nl = "";
+static int is_file = 0;
```
Comment (radare) [inline binr/rax2/rax2.c]: use bool

File: binr/rax2/rax2.c [hunk 2]
```diff
@@ -228,8 +246,43 @@ static int rax(char *str, int len, int last) {
...
+						char octal_str[3], *modified_str;
+						strncpy (octal_str, str + i, 2);
+						i++;
+						modified_str = r_str_newf (((*octal_str != '0') ? "0%s" : "%s"), octal_str);
+						ut64 n = r_num_math (num, modified_str); 
+						printf ("%02x", (ut8) n);
+					} else {	// '\ooo'
+						char octal_str[4], *modified_str;
+						strncpy (octal_str, str + i, 3);
+						i += 2;;
```
Comment (radare) [inline binr/rax2/rax2.c]: double semi colon

File: binr/rax2/rax2.c [hunk 3]
```diff
@@ -228,8 +246,43 @@ static int rax(char *str, int len, int last) {
 		return true;
 	}
 	if (flags & (1 << 2)) { // -S
+		int i, dec_val;
 		for (i = 0; i < len; i++) {
-			printf ("%02x", (ut8) str[i]);
+			if (str[i] == '\\' && !is_file) {
+         		++i;
```
Comment (radare) [inline binr/rax2/rax2.c]: wrong indent and unnecessary  preincrement

**Comment by rahulsangwn** [comment]
done changes...but preincrement  is necessary at line 252

File: binr/rax2/rax2.c [hunk 4]
```diff
@@ -228,8 +246,43 @@ static int rax(char *str, int len, int last) {
 		return true;
 	}
 	if (flags & (1 << 2)) { // -S
+		int i, dec_val;
 		for (i = 0; i < len; i++) {
-			printf ("%02x", (ut8) str[i]);
+			if (str[i] == '\\' && !is_file) {
+         			++i;
```
Comment (radare) [inline binr/rax2/rax2.c]: dont do preincrement here

**Comment by rahulsangwn** [comment]
hey! please let me know fast what is the problem with the increment and how i can eliminate it, so i can work on next issue.

**Comment by rahulsangwn** [comment]
thanks ret2libc for your valuable reply. I had made the changes in `rax()` itself keeping the `is_file` variable.

**Comment by ret2libc** [comment]
> thanks ret2libc for your valuable reply. 

You are welcome!

> I had made the changes in `rax()` itself keeping the `is_file` variable.

Why? Static variables are bad and I still think `is_file` is not necessary in this case.

**Comment by rahulsangwn** [comment]
made the corrections

**Comment by ret2libc** [comment]
I still see `is_file`.

**Comment by rahulsangwn** [comment]
sorry! for previous commit.

File: binr/rax2/rax2.c [hunk 5]
```diff
@@ -228,6 +228,7 @@ static int rax(char *str, int len, int last) {
 		return true;
 	}
 	if (flags & (1 << 2)) { // -S
+		int i;
```
Comment (ret2libc) [inline binr/rax2/rax2.c]: this is not necessary anymore.

**Comment by ret2libc** [comment]
Thanks a lot! This seems great, let's just wait for Travis to test the change.

Also, we usually try to create new tests for new features like the one you added, so that we can ensure the feature is implemented correctly and we don't have regressions in the future. You can see how to create a new test in the radare2-regressions repository (see https://github.com/radare/radare2/blob/master/CONTRIBUTING.md#getting-started and https://github.com/radare/radare2-regressions/blob/master/new/README.md#the-newr2r).

Don't feel forced to do it, though. I just want to mention what we usually do.

**Comment by ret2libc** [comment]
Thanks a lot! I've merged this.

For the future, please use a separate branch for PRs. This way, not only it's better for you in local but it also allows Travis to test your radare2-regressions changes together with your radare2 changes.

**Comment by rahulsangwn** [comment]
thank you so much! r2l, noted all of your points.

**Pull request merged**